PDFExporter v2.0

david-livadaru-3pillar opened the pull request.

Commit:
UITableView

- add support scrollable content, implement UITableView;

Commit:
UITableView

- fixes;
- comment possible unnecessary code;
- add prefix to category names;

Commit:
UITableViewTest

- add test code for UITableView;

Commit:
Refactor-PDFExporter

- remove unnecessary code;
- handle subviews drawing;
- fix issues with state preservation;

Commit:
Refactor-PDFExporter

- fix support for metadata;

Commit:
Refactor-PDFExporter

- fix table view issues;
- remove metadata and paging subview filtering;

Commit:
Refactor-PDFExporter

- update ReadMe file and PDFPrintPageRenderer's header;

Commit:
Non-scaled-text

Add the ability to choose between scaled or non-scaled content.

Commit:
View-slicing

First attempt to implement non slicing views.
Add helper geometry functions.
Add computation for number of pages.

Commit:
View-slicing

Split functionality into separate UIView categories.
Fix dangling pointer for weak reference.
Remove copy of data.
Fix issue of rendering offsets.
Add underlying type for NSMutableArray.
Fix typo for CGPointTranslate.

Commit:
View-slicing

Add helper geometry functions.
Improve page layout splitting.
Fix layout and drawing issues.

Commit:
View-slicing

Remove unnecessary code.

Commit:
View-slicing

Modularize geometry functions.
Fix layout issue with scaled content and non sliced views.

Commit:
View-slicing

Fix issues with FooterView constraints in order to draw it as non scaled view.
Update import headers with framework relative ones.
Fix  drawing issues for UITextView, headers and footers.

Commit:
View-slicing

Fix issues for table view's scroll content.

Commit:
Preparation for version 2.0.

Commit:
Preparation for version 2.0.

Commit:
Preparation for version 2.0.

Commit:
Preparation for version 2.0.

adriansomesan commented:
:+1:

Commit:
Preparation for version 2.0.

vladrusu-3pillar commented on Example/PDFExporterDemo/PDFExporterDemo/Controllers/SettingsScreen/SettingsViewController.h:
```diff
@@ -20,7 +20,7 @@
 @property (weak, nonatomic) id<SettingsViewControllerDelegate> delegate;
 
-@property (nonatomic) UIEdgeInsets contentPaperInsets;
+@property (nonatomic) UIEdgeInsets contentPaperInset;
```
could be a typo here, as the type is UIEdgeInsets and the name of the property is Inset

david-livadaru-3pillar replied:
As per Apple naming convention, we use the singular form (e.g.: contentInset from `UIScrollView`, separatorInset from `UITableViewCell`).

vladrusu-3pillar commented on Framework/PDFExporter/PDFExporter/Categories/UIKit/UITextView+PDFExporterDrawing.m:
```diff
@@ -0,0 +1,16 @@
+
+@implementation UITextView (PDFExporterDrawing)
+
+- (BOOL)canDrawSubview:(UIView *)subview intersection:(CGRect)intersection {
```
Would be nice if cases like this one were documented, it's hard to understand what's happening

david-livadaru-3pillar replied:
We have written documentation for source code.

vladrusu-3pillar commented:
:+1:

sean7512 commented on Framework/PDFExporter/PDFExporter/Categories/UIKit/UIScrollView+PDFExporterDrawing.m:
```diff
@@ -0,0 +1,85 @@
+#pragma mark - PDFExporterPageInformation
+
+- (CGRect)subviewRect:(UIView *)subview layoutPageRect:(CGRect)rect {
+    if ([subview isKindOfClass:[UITableViewCell class]]) {  // UITableView's cells are wrapped inside a UIScrollView
```
this is a big assumption and Apple can change this "expected" behavior at any time.

You should check if superview exists and is a scrollview instead?

FYI: I am a US 3Pillar employee

gabrieldusa replied:
This is a good idea. I also think we should add this check.

david-livadaru-3pillar replied:
subview.superview is self in this case. The check is redundant.

sean7512 commented on Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m:
```diff
@@ -0,0 +1,125 @@
+    CGPoint offset = CGPointZero;
+    for (CGFloat yCoordinate = CGRectGetMinY(rect);
+         yCoordinate < CGRectGetMaxY(rect);
+         yCoordinate += CGRectGetHeight(self.bounds)) {
```
No need to call CGRectGetMaxY and CGRectGetHeight for each iteration

david-livadaru-3pillar replied:
What is happening here is that we're looping from minimum Y to maximum Y with using the height of bounds as step.

sean7512 commented on Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m:
```diff
@@ -0,0 +1,125 @@
+        self.contentOffset = offset;
+        CGRect frame = self.frame;
+        [self layoutIfNeeded];
+        self.frame = frame;
```
Do you expect `self.frame` to change after layout?  If it does change, it ends up being overridden back to the original frame anyways.  Why is this necessary?

david-livadaru-3pillar replied:
It is not needed anymore. It has been removed.

sean7512 commented on Framework/PDFExporter/PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m:
```diff
@@ -0,0 +1,125 @@
+        CGRect frame = self.frame;
+        [self layoutIfNeeded];
+        self.frame = frame;
+        [self layoutHeadersAndFooters];
```
this chunk of code is duplicated and should be pulled out

david-livadaru-3pillar replied:
We have removed the duplicated code.

sean7512 commented on Framework/PDFExporter/PDFExporter/Categories/UIKit/UIView+PDFExporterStatePersistance.m:
```diff
@@ -0,0 +1,86 @@
+
+@implementation UIView (PDFExporterStatePersistance)
+
+- (NSMutableArray<UIViewPersistenceState *> *)persistenceStates {
```
:+1: For using generics...it would be nice to use them everywhere as well.  Makes Swift use a lot easier

sean7512 commented:
Would be nice to see the API audited for Nullability for easier use in Swift

david-livadaru-3pillar committed:
Merge branch 'master' into PDFExporter_v2.0

# Conflicts:
#	Framework/PDFExporter/PDFExporter.xcodeproj/project.pbxproj

david-livadaru-3pillar committed:
Add support for page orientation.

david-livadaru-3pillar committed:
Add framework.build folder in gitignore.
Remove custom script from build settings (there is a target available for universal framework).

david-livadaru-3pillar committed:
Perform layout before any geometry computation.
Fix number of pages compution when views are sliced.

david-livadaru-3pillar committed:
Add a UIView subview which is able to draw images from .pdf files.

david-livadaru-3pillar committed:
Fix missing cell from on page break from PDF when views are not sliced.

david-livadaru-3pillar committed:
Make sure that layout is performed on main queue.
Perform cleanup correctly when PDF drawing is cancelled in the middle drawing.

david-livadaru-3pillar committed:
Fix global function import linkage issue.

david-livadaru-3pillar committed:
Perform all table view layout in the same block on main queue.

david-livadaru-3pillar committed:
Update Build Active Architectures for Release configuration.

BogdanTC committed:
origin/PDFExporter_v2.0

Removed demo app from framework, ready to use with Carthage or Cocoapods.

BogdanTC committed:
origin/PDFExporter_v2.0

Fixed podspec typo.

BogdanTC committed:
origin/PDFExporter_v2.0

Fixed CocoaPods error.

BogdanTC committed:
origin/PDFExporter_v2.0

Readme text updated.

ciprian-antal-3PG committed:
Remove duplicated code.

david-livadaru-3pillar committed:
table-view-frame-preservation

Remove unnecessary frame preservation for table view while performing layout.
Save state for rendering offset.
Perform persitance preparation before computing anything.

ciprian-antal-3PG committed:
Update with PR comments.
Rename function.

ciprian-antal-3PG committed:
Rename 'setFrameForContentOffset' method to 'updateContentOffset'.

david-livadaru-3pillar committed:
layout-fixes

Add support for comparing CGFloat values.
Add default YES value for drawEntireContentSize from UIScrollView and persistState from UIView.
Save state of UITableView while drawing.
Use CGFloatIsEqual instead of plain == operator.

david-livadaru-3pillar committed:
documentation

Add documentation for HeaderFooterPaging, PaperSize and rendering delegate.

david-livadaru-3pillar committed:
layout-fixes

Change default of drawEntireContentSize and persistState to NO.

david-livadaru-3pillar committed:
Merge branch 'layout-fixes' into documentation

david-livadaru-3pillar committed:
documentation

Add extension as public headers.
Split UIView's extension into separate Objective-C files.
Create separate files for state classes.
Add documentation for the rest of files.
Remove unused paramter size from drawing preparation.
Use CGMake function instead of creating temporary objects in convenince functions.
Declare protocols and classes in headers and import the headers in the implementation files.

david-livadaru-3pillar committed:
Merge pull request #6 from 3pillarlabs/feature/Remove-duplicated-code

Remove duplicated code.

david-livadaru-3pillar committed:
Merge branch 'PDFExporter_v2.0' into layout-fixes

# Conflicts:
#	PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m

ciprian-antal-3PG committed:
Merge pull request #8 from 3pillarlabs/layout-fixes

Layout fixes

david-livadaru-3pillar committed:
Merge branch 'PDFExporter_v2.0' into table-view-frame-preservation

# Conflicts:
#	PDFExporter/Categories/UIKit/UITableView+PDFExporterDrawing.m

ciprian-antal-3PG committed:
Merge pull request #7 from 3pillarlabs/table-view-frame-preservation

TableView frame preservation

david-livadaru-3pillar committed:
Merge branch 'PDFExporter_v2.0' into documentation

# Conflicts:
#	PDFExporter/PDFRender/PDFPrintPageRenderer.m

david-livadaru-3pillar committed:
documentation

Remove unused setters.

ciprian-antal-3PG committed:
Merge pull request #9 from 3pillarlabs/documentation

Documentation

david-livadaru-3pillar committed:
PDFExporter_v2.0

Add missing header import in framework.

david-livadaru-3pillar committed:
PDFExporter_v2.0

Remove unusued memory cleaner.
Fix floating points equality.

ciprian-antal-3PG committed:
remove-drawing-on-background-thread

Remove drawing on background thread.

david-livadaru-3pillar committed:
Merge pull request #10 from 3pillarlabs/remove--drawing-on-background-thread

Remove drawing on background thread.
